File: 0310-minimum-height-trees/0310-minimum-height-trees.py
class Solution:
    def dfs(self, src: int, adj: List[List[int]], prev: int) -> List[int]:
        if len(adj[src]) == 1 and prev == adj[src][0]: return [0, src]
        dist = 0
        e = -1
        for edge in adj[src]:
            if edge != prev:
                t = self.dfs(edge, adj, src)
                t[0] += 1
                if t[0] > dist:
                    dist = t[0]
                    e = t[1]
        return [dist, e]
    path = []
    def dfs2(self, src: int, adj: List[List[int]], prev: int, end: int) -> bool:
        if len(adj[src]) == 1 and prev == adj[src][0]:
            if src == end:
                self.path.append(end)
                return True
            else:
                return False
        for edge in adj[src]:
            if edge != prev:
                if self.dfs2(edge, adj, src, end):
                    self.path.append(src)
                    return True
        return False
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n <= 2:
            ans = [i for i in range(n)]
            return ans
        adj = [[] for _ in range(n)]
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])
        start = self.dfs(0, adj, -1)[1]
        end = self.dfs(start, adj, -1)[1]
        self.path = []
        self.dfs2(start, adj, -1, end)
        pathLen = len(self.path)
        if pathLen % 2 == 0:
            return [self.path[pathLen//2], self.path[pathLen//2-1]]
        else:
            return [self.path[pathLen//2]]

File: 0310-minimum-height-trees/0310-minimum-height-trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if(n <= 2) {
            var ans = new LinkedList<Integer>();
            for(int i=0; i<n; i++) ans.add(i);
            return ans;
        }
        var adj = new ArrayList<Set<Integer>>();
        for(int i=0; i<n; i++) adj.add(new HashSet<Integer>());
        for(var edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }
        var queue = new LinkedList<Integer>();
        for(int i=0; i<n; i++) if(adj.get(i).size() == 1) queue.add(i);
        while(true) {
            int size = queue.size();
            var prevQueue = new LinkedList<Integer>();
            while(size-- > 0) {
                var t = queue.poll();
                prevQueue.add(t);
                if(adj.get(t).size() == 0) continue;
                var t1 = adj.get(t).iterator().next();
                adj.get(t1).remove(t);
                if(adj.get(t1).size() == 1) queue.add(t1);
            }
            if(queue.isEmpty()) return prevQueue;
        }
    }
}